File: 39-combination-sum/combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        def combinations(sub, i, target):
            if i==len(candidates):
                if target==0:
                    ans.append(sub)
                return

            if candidates[i]<=target:
                cub = sub.copy()
                cub.append(candidates[i])
                combinations(cub, i,target-candidates[i])
                
            combinations(sub, i+1, target)

        combinations([], 0, target)
        return ans

File: 39-combination-sum/combination-sum.java
class Solution {
    List<List<Integer>> result = new ArrayList<>();

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        backtrack(candidates, target, 0, new ArrayList<>());
        return result;
    }

    private void backtrack(int[] candidates, int target, int start, List<Integer> path) {
        if (target == 0) {
            result.add(new ArrayList<>(path));
            return;
        }

        if (target < 0) return;

        for (int i = start; i < candidates.length; i++) {
            path.add(candidates[i]);
            backtrack(candidates, target - candidates[i], i, path); // reuse allowed
            path.remove(path.size() - 1); // backtrack
        }
    }
}